src/backend/ctrl_flow_struct/mod.rs

```diff
@@ -0,0 +1,466 @@
+        (ret, slice_topo_order)
+    }
+
+    /// Transforms the loop into a single-entry loop.
+    /// Returns the new loop header.
+    fn funnel_abnormal_entries(&mut self, header: NodeIndex, loop_nodes: &NodeSet) -> NodeIndex {
+        if header == self.entry {
+            // all entries must go through `entry` aka `header`
+            return header;
+        }
+
+        let mut entry_map = HashMap::new();
+        for n in loop_nodes {
+            for e in self.graph.edges_directed(n, Incoming) {
+                if !loop_nodes.contains(e.source()) {
+                    entry_map.entry(n).or_insert(Vec::new()).push(e.id());
+                }
+            }
+        }
+        println!("entries: {:?}", entry_map);
+        // loop must be reachable, so the header must have entries
+        let header_entries = entry_map.remove(&header).unwrap();
+        debug_assert!(!header_entries.is_empty());
+        let abnormal_entry_map = entry_map;
+        if abnormal_entry_map.is_empty() {
+            // no abnormal entries
+            return header;
+        }
+        let abnormal_entry_iter = (1..).zip(&abnormal_entry_map);
+
+        let struct_var = "i".to_owned(); // XXX
+
+        // make condition cascade
+        let new_header = {
+            let abnormal_entry_iter = abnormal_entry_iter.clone().map(|(n, (&t, _))| (n, t));
+
+            let dummy_preheader = self.graph.add_node(CfgNode::Dummy("loop \"preheader\""));
+
+            let mut prev_cascade_node = dummy_preheader;
+            let mut prev_entry_target = header;
+            let mut prev_out_cond = None;
+            let mut prev_entry_num = 0;
+
+            // we make the condition node for the *previous* entry target b/c
+            // the current one might be the last one, which shouldn't get a
+            // condition node because it's the only possible target
+            for (entry_num, entry_target) in abnormal_entry_iter {
+                let cascade_node = self.graph.add_node(CfgNode::Condition);
+                self.graph
+                    .add_edge(prev_cascade_node, cascade_node, prev_out_cond);
+                self.graph.add_edge(
+                    cascade_node,
+                    prev_entry_target,
+                    Some(self.cctx.mk_simple(SimpleCondition(format!(
+                        "{} == {}",
+                        struct_var, prev_entry_num
+                    )))),
+                ); // XXX
+                let struct_reset = self.graph.add_node(CfgNode::Code(AstNode::BasicBlock(
+                    format!("{} = 0", struct_var),
+                ))); // XXX
+                self.graph.add_edge(struct_reset, entry_target, None);
+                prev_cascade_node = cascade_node;
+                prev_entry_target = struct_reset;
+                prev_out_cond = Some(self.cctx.mk_simple(SimpleCondition(format!(
+                    "{} != {}",
+                    struct_var, prev_entry_num
+                )))); // XXX
```

Comment:
What are these `XXX` for? If its something incomplete or a todo, please include a comment. Similar ones a few lines above.

Reply:
The `XXX` means it needs to be replaced with actual values from `SimpleCAST` when that gets merged.